lib/mirage/mirage.mli

```diff
@@ -764,6 +764,89 @@ val default_argv : argv impl
 val no_argv : argv impl
 (** [no_argv] Disable command line parsing and set argv to [|""|]. *)
 
+(** {2 Mimic device} *)
+
+(** A {i mimic} device is a {i context} which contains few informations to let
+    the user to [connect]/{!Mimic.resolve}/allocate a new resource to
+    communicate with a peer. Depending on what the context contains, the
+    protocol used to communicate can be TCP/IP, TLS or SSH (or something else).
+
+    Few packages deliver a {i context} which allows the user to communicate to a
+    server via a {i certain} protocol. The user is able to {i merge} these
+    provided contexts to be able to communicate with a server with multiple
+    protocols. Then, the choice is made dynamically according to what the
+    context contains. *)
+
+type mimic
+
+val mimic : mimic typ
+(** A simple device which define a {i certain} Mimic's context. *)
+
+val mimic_merge : mimic impl -> mimic impl -> mimic impl
+(** [mimic_merge ctx0 ctx1] merges given contexts into one. *)
+
+(** {2 Git configuration} *)
+
+(** The user is able to initiate a connection with a Git repository. However, it
+    exists multiple ways to communicate with a Git repository:
+
+    - TCP/IP
+    - HTTP
+    - HTTP + TLS
+    - SSH
+
+    These devices define these composable ways and deliver a Mimic's context
+    which can be pass to Git or Irmin in order to be able to {i fetch} and
+    {i push} from/into a Git repository. *)
+
+val git_happy_eyeballs :
+  ?random:random impl ->
+  ?time:time impl ->
+  ?mclock:mclock impl ->
+  ?pclock:pclock impl ->
+  stackv4v6 impl ->
+  mimic impl
+(** [git_happy_eyeballs stackv4v6] provides a {!mimic} device which is {b only}
+    able to resolve domain-name. The user must {!mimic_merge} this device with
+    another which provides a context with a {i certain} protocol. *)
+
+val git_tcp : tcpv4v6 impl -> mimic impl -> mimic impl
+(** [git_tcp tcpv4v6 ctx] returns a new {!mimic} context which is able to
+    allocate a TCP/IP connection with a remote Git repository. *)
+
+val git_ssh :
+  ?authenticator:string key ->
+  key:string key ->
+  ?mclock:mclock impl ->
+  ?time:time impl ->
+  tcpv4v6 impl ->
+  mimic impl ->
+  mimic impl
+(** [git_ssh ?authenticator ~key tcpv4v6 ctx] returns a new {!mimic} context
+    which is able to allocate an SSH connection with the given private [key]
+    with a remote Git repository. You can ensure the identity of the remote Git
+    repository via [authenticator].
+
+    The format of the private key is: [<type>:<base64>] The format of the
+    authenticator is: [<domain>:<hash>:<fingerprint>] *)
+
+val git_http :
+  ?tls_key_fingerprint:string key ->
+  ?tls_cert_fingerprint:string key ->
+  ?headers:(string * string) list key ->
+  ?time:time impl ->
+  ?pclock:pclock impl ->
+  tcpv4v6 impl ->
+  mimic impl ->
+  mimic impl
+(** [git_http ?tls_key_fingerprint ?tls_cert_fingerprint ?headers tcpv4v6 ctx]
```

Comment:
I still find it kind of strange what authenticator is used:
- by default a chain of trust with ca-certs-nss
- if tls_key_fingerprint is provided, this value (unclear which format this should be in - likely hex?)
- if tls_cert_fingerprint is provided, that value

is it possible to provide a custom trust anchor? it should be. what will happen if both tls_key_fingerprint and tls_cert_fingerprint are provided? is there still a possibility to have no validation?

so. similar to the ssh above, I'd be in favour of a single `authenticator` (optional) labeled argument, and its value should include:
- key;hash?:fingerprint
- cert:hash?:fingerprint
- trust-anchor:pem-encoded cert (list?)
- none

Reply:
Yeah we can improve that but about `trust-anchor:pem-encoded cert`, you mean a file (which must be read) or a serialized form? And if we have `none`, that means Mozilla's NSS trust anchors?

Reply:
> you mean a file

no, just the raw pem-encoded certificate (if argument length allows that - at least EC certificates should fit nicely into the limit). a file would be rather complex to support

> `none`, that means Mozilla's NSS trust anchors?

no, it should mean none --> no, thanks, no validation whatsoever should be done. mainly for debugging purposes, or for scenarios where you don't care.

Reply:
but how you want then to handle Mozilla's NSS trust anchors?

Reply:
as default? also an explicit `nss-trust-anchors` value for `authenticator` could be supported which uses the nss trust anchors. My main arguments are:
- why have multiple optional keyword arguments, and what is the semantics if all are present?
- there are at least two use cases not supported:
   - a custom trust anchor (self-signed certificate in environments where you have your own CA), and 
   - "no validation of the remote server certificate at all".